### DISCUSSION TITLE
How can I use report intermediate result about multi metrics?

### DISCUSSION THREAD

**Discussion by rickyyyxu** [microsoft/nni#4289] [Q&A] [answered]
How can I use report intermediate result about **multi metrics**?
For example, I want to report loss, train_loss and test_loss.
Thanks a lot

**Environment**:
- NNI version:
- Training service (local|remote|pai|aml|etc):
- Client OS:
- Server OS (for remote mode only): Ubuntu 16.04.6 LTS
- Python version: 
- PyTorch/TensorFlow version:
- Is conda/virtualenv/venv used?: Yes
- Is running in Docker?: No

**Configuration**:
 - Experiment config (remember to remove secrets!):
 - Search space:

**Log message**:
 - nnimanager.log:
 - dispatcher.log:
 - nnictl stdout and stderr:
 

**How to reproduce it?**:

**Comment by Lijiaoa** [comment]
> nni.report_intermediate_result({'default': 1, 'train_loss': xxx, 'test_loss': xxx})

It's helpful for you? Please let me know. @coasxu

**Comment by rickyyyxu** [comment]
> > nni.report_intermediate_result({'default': 1, 'train_loss': xxx, 'test_loss': xxx})
> 
> It's helpful for you? Please let me know. @coasxu

Thank you for your reply!
By `dict`, I can compare another metric's change in a whole training process, like this.
![image](https://user-images.githubusercontent.com/24477758/139267641-9a9b2fcc-1839-4b42-8d6a-6a7aaab4c756.png)

But, how can nni compare several metrics between different trials. Here only show the default metric.

![image](https://user-images.githubusercontent.com/24477758/139269614-b0b985a5-6677-413d-8b73-f9f739de930b.png)

**Reply by louiswng** [reply]
I have the same question.

**Reply by Lijiaoa** [reply]
@louiswong-cs This is the answer: https://github.com/microsoft/nni/discussions/4289#discussioncomment-1579727

**Comment by Lijiaoa** [comment, marked as answer]
You could choose some trials in the table and then click the button named `Compare`. `Compare` model  is designed for this function. Look like this picture:

![image](https://user-images.githubusercontent.com/61399850/139362739-6412db4a-0304-4a19-8f93-8225dbff8a23.png)

**Reply by Annika10** [reply]
Hi @Lijiaoa, I cannot find the drop down menu, also not for the intermediate results even by adding multiple metrics like you suggested above. It is also not described where it could be found in the [tutorial](https://nni.readthedocs.io/en/stable/Tutorial/WebUI.html?highlight=webui). Is it still present in the WebUI?

**Reply by Lijiaoa** [reply]
> Hi @Lijiaoa, I cannot find the drop down menu, also not for the intermediate results even by adding multiple metrics like you suggested above. It is also not described where it could be found in the [tutorial](https://nni.readthedocs.io/en/stable/Tutorial/WebUI.html?highlight=webui). Is it still present in the WebUI?

Hi, @Annika10 
WebUI doesn't supports this function on version `v1.9-2.2`. So what's your version? Could you use the latest version to check it? And you could use the API `http://your-ip:your-port/api/v1/nni/metric-data` to check your dict intermediate results.  Please let me know if you have any problem.

**Reply by Annika10** [reply]
> > Hi @Lijiaoa, I cannot find the drop down menu, also not for the intermediate results even by adding multiple metrics like you suggested above. It is also not described where it could be found in the [tutorial](https://nni.readthedocs.io/en/stable/Tutorial/WebUI.html?highlight=webui). Is it still present in the WebUI?
> 
> Hi, @Annika10 WebUI doesn't supports this function on version `v1.9-2.2`. So what's your version? Could you use the latest version to check it? And you could use the API `http://your-ip:your-port/api/v1/nni/metric-data` to check your dict intermediate results. Please let me know if you have any problem.

I use nni version 2.5. I just have the filter option like it is explained in the [tutorial](https://nni.readthedocs.io/en/stable/Tutorial/WebUI.html?highlight=webui#view-trial-intermediate-result-graph) or are the intermediate results also somewhere else showed in the Web UI? The links works though but I would like to have the graph

**Reply by Lijiaoa** [reply]
ok. I know that. 
In fact you should use the `Compare` model to see the trials intermediate results. Firstly you could select some trials you want to see and then click the button named `Compare`. And we don't support this function in the all trials intermediate results graph. Because `Compare model` is for that. Expect your reply. @Annika10 

![compare](https://user-images.githubusercontent.com/61399850/148152595-c045124f-0a54-4696-ba68-2df9090c5187.png)

**Reply by Annika10** [reply]
Excuse the circumstance, but I cannot find the drop down menu.
![Screenshot from 2022-01-05 10-07-46](https://user-images.githubusercontent.com/28507653/148191137-4f92d3d6-6191-4af7-b95f-935d75214992.png)
In my pytorch lightning module my code looks like this:
```
    def on_validation_epoch_end(self):
        nni.report_intermediate_result({'default': self._get_validation_metrics(), 'val_loss': self.val_loss})
```

**Reply by Lijiaoa** [reply]
So what's your api `http://your-ip:your-port/api/v1/nni/metric-data` data?

**Reply by Annika10** [reply]
Ah you're right! My own `val_loss` metric was a tensor. I converted it to float and now it works! Thanks a lot and sorry for all the circumstances :)

**Reply by Lijiaoa** [reply]
Congratulations!

**Reply by jaemin93** [reply]
I changed to float and reported to dict type, but the drop down menu doesn't appear.
```
[{"timestamp":1651619142597,"trialJobId":"NjUDN","parameterId":"0","type":"PERIODICAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619156155,"trialJobId":"NjUDN","parameterId":"0","type":"PERIODICAL","sequence":1,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619159108,"trialJobId":"NjUDN","parameterId":"0","type":"FINAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619197631,"trialJobId":"NIaa1","parameterId":"1","type":"PERIODICAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619211311,"trialJobId":"NIaa1","parameterId":"1","type":"PERIODICAL","sequence":1,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619214324,"trialJobId":"NIaa1","parameterId":"1","type":"FINAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619263113,"trialJobId":"NV6tZ","parameterId":"2","type":"PERIODICAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619286527,"trialJobId":"NV6tZ","parameterId":"2","type":"PERIODICAL","sequence":1,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619289730,"trialJobId":"NV6tZ","parameterId":"2","type":"FINAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619338383,"trialJobId":"tlFQL","parameterId":"3","type":"PERIODICAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619361898,"trialJobId":"tlFQL","parameterId":"3","type":"PERIODICAL","sequence":1,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""},{"timestamp":1651619365156,"trialJobId":"tlFQL","parameterId":"3","type":"FINAL","sequence":0,"data":"\"{\\\"default\\\": 0.5, \\\"mae\\\": 0.3, \\\"acc\\\": 0.8}\""}]
```

![image](https://user-images.githubusercontent.com/29832101/166588151-e4e9e82f-3592-4b34-8329-878fb794f15e.png)

**Reply by Lijiaoa** [reply]
@jaemin93 Could you use nni latest version to check it? 'dict' is not supported in `v1.9-v2.2`... Please let me know if you still have any problem.

**Reply by jaemin93** [reply]
> @jaemin93 Could you use nni latest version to check it? 'dict' is not supported in `v1.9-v2.2`... Please let me know if you still have any problem.

I'm using the 2.7 version. :)

**Reply by Lijiaoa** [reply]
yes. will fix it in v2.8

**Reply by Lijiaoa** [reply]
@jaemin93 if you want to use compare function in current, you could click `intermediate button` first. The operation could avoid this bug.
